### ISSUE TITLE
Errors when LoadBalancer is not deployed?

### ISSUE THREAD

**Issue by alexellis** [equinix/cloud-provider-equinix-metal#115]
The README says that the LoadBalancer is optional, however when I deploy the CCM without the LoadBalancer I see a lot of errors in the CCM logs asking to edit MetalLB resources.

Full logs are provided below.

What would you suggest?

```
root@ccm-test-1:~# RELEASE=v1.1.0
root@ccm-test-1:~# kubectl apply -f https://github.com/packethost/packet-ccm/releases/download/${RELEASE}/deployment.yaml
deployment.apps/packet-cloud-controller-manager created
serviceaccount/cloud-controller-manager created
clusterrole.rbac.authorization.k8s.io/system:cloud-controller-manager created
clusterrolebinding.rbac.authorization.k8s.io/system:cloud-controller-manager created

root@ccm-test-1:~# kubectl logs deployment.apps/packet-cloud-controller-manager -n kube-system
Flag --allow-untagged-cloud has been deprecated, This flag is deprecated and will be removed in a future release. A cluster-id will be required on cloud instances.
Flag --allow-untagged-cloud has been deprecated, This flag is deprecated and will be removed in a future release. A cluster-id will be required on cloud instances.
Flag --allow-untagged-cloud has been deprecated, This flag is deprecated and will be removed in a future release. A cluster-id will be required on cloud instances.
I1203 14:39:04.122130       1 serving.go:312] Generated self-signed cert in-memory
W1203 14:39:06.168099       1 client_config.go:543] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1203 14:39:06.174304       1 controllermanager.go:120] Version: v0.0.0-master+$Format:%h$
W1203 14:39:06.174352       1 controllermanager.go:132] detected a cluster without a ClusterID.  A ClusterID will be required in the future.  Please tag your cluster to avoid any future issues
I1203 14:39:06.176071       1 secure_serving.go:178] Serving securely on [::]:10258
I1203 14:39:06.176213       1 tlsconfig.go:219] Starting DynamicServingCertificateController
2020/12/03 14:39:06 [DEBUG] POST https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/bgp-configs
E1203 14:39:06.791637       1 cloud.go:210] failed to update and sync nodes for handler: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:39:06.791672       1 cloud.go:210] failed to update and sync nodes for handler: elastic ip tag is empty. Nothing to do
I1203 14:39:06.791773       1 cloud.go:237] nodes watcher started
2020/12/03 14:39:06 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:39:06.802529       1 cloud.go:222] failed to update and sync node for add ccm-test-1 for handler: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:39:06.802605       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:39:06.802649       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:39:06.802682       1 cloud.go:222] failed to update and sync node for add ccm-test-1 for handler: elastic ip tag is empty. Nothing to do
E1203 14:39:07.353793       1 cloud.go:260] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
I1203 14:39:07.353895       1 cloud.go:285] services watcher started
I1203 14:39:07.353934       1 core.go:101] Will not configure cloud provider routes for allocate-node-cidrs: false, configure-cloud-routes: true.
W1203 14:39:07.353977       1 controllermanager.go:244] Skipping "route"
I1203 14:39:07.357382       1 node_controller.go:110] Sending events to api server.
I1203 14:39:07.357568       1 controllermanager.go:247] Started "cloud-node"
2020/12/03 14:39:07 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:39:07.360403       1 node_controller.go:237] provider name from providerID should be packet, was k3s
2020/12/03 14:39:07 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/devices?include=facility
I1203 14:39:07.360804       1 node_lifecycle_controller.go:77] Sending events to api server
I1203 14:39:07.360885       1 controllermanager.go:247] Started "cloud-node-lifecycle"
E1203 14:39:07.364310       1 core.go:90] Failed to start service controller: the cloud provider does not support external load balancers
W1203 14:39:07.364347       1 controllermanager.go:244] Skipping "service"
E1203 14:39:08.260536       1 cloud.go:271] failed to update and sync service for add kube-system/metrics-server: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
2020/12/03 14:39:08 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:39:08.927122       1 cloud.go:271] failed to update and sync service for add kube-system/traefik-prometheus: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
2020/12/03 14:39:08 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:39:09.630507       1 cloud.go:271] failed to update and sync service for add kube-system/traefik: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
2020/12/03 14:39:09 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:39:10.133549       1 cloud.go:271] failed to update and sync service for add default/kubernetes: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
2020/12/03 14:39:10 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:39:10.733212       1 cloud.go:271] failed to update and sync service for add kube-system/kube-dns: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
2020/12/03 14:40:07 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:40:07.930101       1 cloud.go:302] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:40:07.931886       1 cloud.go:311] failed to update and sync nodes: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:40:07.931976       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:40:07.932028       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:40:07.932059       1 cloud.go:311] failed to update and sync nodes: elastic ip tag is empty. Nothing to do
2020/12/03 14:41:07 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:41:08.629716       1 cloud.go:302] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:41:08.631329       1 cloud.go:311] failed to update and sync nodes: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:41:08.631425       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:41:08.631481       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:41:08.631507       1 cloud.go:311] failed to update and sync nodes: elastic ip tag is empty. Nothing to do
2020/12/03 14:42:08 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:42:09.276806       1 cloud.go:302] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:42:09.278411       1 cloud.go:311] failed to update and sync nodes: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:42:09.278514       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:42:09.278566       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:42:09.278596       1 cloud.go:311] failed to update and sync nodes: elastic ip tag is empty. Nothing to do
2020/12/03 14:43:09 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:43:10.005317       1 cloud.go:302] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:43:10.007068       1 cloud.go:311] failed to update and sync nodes: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:43:10.007162       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:43:10.007212       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:43:10.007255       1 cloud.go:311] failed to update and sync nodes: elastic ip tag is empty. Nothing to do
E1203 14:44:09.083326       1 node_controller.go:237] provider name from providerID should be packet, was k3s
2020/12/03 14:44:09 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/devices?include=facility
2020/12/03 14:44:10 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:44:10.837322       1 cloud.go:302] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:44:10.839266       1 cloud.go:311] failed to update and sync nodes: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:44:10.839357       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:44:10.839415       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:44:10.839462       1 cloud.go:311] failed to update and sync nodes: elastic ip tag is empty. Nothing to do
2020/12/03 14:45:10 [DEBUG] GET https://api.packet.net/projects/66ae0069-7d03-4db5-9af1-6b14036e380a/ips?
E1203 14:45:11.829765       1 cloud.go:302] failed to update and sync services: unable to retrieve metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:45:11.831526       1 cloud.go:311] failed to update and sync nodes: failed to get metallb config map: unable to get metallb configmap config: configmaps "config" is forbidden: User "system:serviceaccount:kube-system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "metallb-system"
E1203 14:45:11.831622       1 bgp.go:79] could not ensure BGP enabled for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:45:11.831678       1 bgp.go:88] bgp.reconcileNodes(): could not get BGP info for node ccm-test-1: provider name from providerID should be packet, was k3s
E1203 14:45:11.831719       1 cloud.go:311] failed to update and sync nodes: elastic ip tag is empty. Nothing to do

```

**Comment by alexellis** [comment]
@deitch these errors are still present in v2.0.0 - but we see fewer repeated errors.

**Comment by deitch** [comment]
This is more of a documentation problem. It tells you if metallb is there, so that if it fails, you know why. We could make those debug?

**Comment by alexellis** [comment]
I wouldn't expect the code in the CCM to try to configure and access MetalLB resources, if it's deployed without an LB present. I don't know if that would be a flag, or some kind of auto-detection.

Perhaps the code that tries to get the configmap should do that once upon creation and then use the result to decide whether it should do it again or not? That way it would print just the one "error" code.

**Comment by deitch** [comment]
No, it should be checking all the time. You shouldn't have to kill and restart the ccm to enable metallb. The bigger question is how to report that state without making it look like an error.

**Comment by displague** [comment]
@deitch 
`unable to get metallb configmap config: configmaps "config" is forbidden:` were these RBAC issues?
"Forbidden" is not the response I would expect if the configmap did not exist, unless that is a result of the namespace not existing, and perhaps that is the case.

The minutely check for the configmap seems excessive, would it poll less frequently if the configmap did exist?

@alexellis 
> The README says that the LoadBalancer is optional

It is optional but enabled by default. Set `PACKET_LB_CONFIGMAP=disabled` in the environment to avoid all of this.

**Comment by deitch** [comment]
> The minutely check for the configmap seems excessive, would it poll less frequently if the configmap did exist?

Actually, I would like it to check more frequently. We don't want the metallb configmap to be a prerequisite, such that if it isn't there on ccm start, we would have to restart the ccm. We want it to be able to pick up the configmap even if it is created 1 second (or 1 week) afterwards.

> It is optional but enabled by default. Set PACKET_LB_CONFIGMAP=disabled in the environment to avoid all of this.

@alexellis, as of now, metallb is "the blessed" load balancer for EM. You don't _have_ to use it, but either you are enabling it (default) or you are not (set the env var that @displague referenced above).

We are having a discussion in another issue #117 about how to broaden it and make the UX of controlling it more flexible and more sane. Feel free to voice an opinion there.

**Comment by deitch** [comment]
> unable to get metallb configmap config: configmaps "config" is forbidden: were these RBAC issues?

It sounds like it, doesn't it @displague ? But as far as I know, it is not, it is because the namespace is not there. The permissions for the ccm are [here](https://github.com/packethost/packet-ccm/blob/master/deploy/template/deployment.yaml#L121-L129)

```yml
  resources:
  - configmaps
  verbs:
  - create
  - get
  - list
  - watch
  - update
  - patch
```

That is a `ClusterRole`, so unless someone messed with it, that should be able to do everything (except for delete) to any configmap in any namespace.

**Comment by thebsdbox** [comment]
With the changes that have been merged, this isn't a relevant behaviour any more?

**Comment by deitch** [comment]
These are pretty much fixed since #220 ; if there remain issues, we should open a new issue.
